File: 12865.cpp
#include <iostream>
#include <algorithm>
#include <climits>
using namespace std;


int dp[100000];
int numItem;


int main() {
    int weight, value, maxWeight;
    cin >> numItem >> maxWeight;
    
    dp[0] = 0;
    
    for(int i = 0; i < numItem; i++) {
        cin >> weight >> value;
        for(int i = maxWeight; i > 0; i--) {
            if (i - weight >= 0)
                dp[i] = max(dp[i], dp[i - weight] + value);
        }
    }
    
    cout << dp[maxWeight];
}

File: 12865.py
import sys

input = lambda: sys.stdin.readline().strip()

num_stocks, max_weight = map(int, input().split())

stocks = [tuple(map(int, input().split())) for i in range(num_stocks)]
stocks = [0] + stocks
dp = [[0] * (max_weight + 1) for i in range(num_stocks + 1)]

for i in range(1, num_stocks + 1):
    for w in range(1, max_weight + 1):
        weight, value = stocks[i]

        if w < weight:
            dp[i][w] = dp[i - 1][w]
        else:
            dp[i][w] = max(dp[i - 1][w], dp[i - 1][w - weight] + value)


print(dp[num_stocks][max_weight])
